### PR TITLE
Add support for TLS imports/exports to/from DSOs

### PR DESCRIPTION
**Pull request by sbc100**
This change is based on the recently added EXPORT_INFO metadata that tells
the dynamic linker which exports are TLS exports and which are normal data
exports: https://reviews.llvm.org/D108877

### CODE REVIEW

**Comment by sbc100** [comment]
All tests now passing.  ptal

File: src/library_pthread.js [hunk 1]
```diff
@@ -490,6 +490,34 @@ var LibraryPThread = {
...
+    // according to this new __tls_base.
+    function tlsInitWrapper() {
+      var __tls_base = tlsInitFunc();
+#if DYLINK_DEBUG
+      err('tlsInit -> ' + __tls_base);
+#endif
+      for (var sym in metadata.tlsExports) {
+        metadata.tlsExports[sym] = moduleExports[sym];
+      }
+      relocateExports(metadata.tlsExports, __tls_base, true);
```
Comment (kripken) [inline src/library_pthread.js:509]:
```suggestion
      relocateExports(metadata.tlsExports, __tls_base, /* replace= */ true);
```

File: src/library_pthread.js [hunk 2]
```diff
@@ -490,6 +490,34 @@ var LibraryPThread = {
...
+    }
+
+    PThread.tlsInitFunctions.push(tlsInitWrapper);
+    if (runtimeInitialized) {
+      tlsInitWrapper();
+    }
```
Comment (kripken) [inline src/library_pthread.js:515]: do we need to both push this function and call it now, if the runtime is initialized? If it's *not* initialized, when would it get called - just at the time the pushed function is?

**Review by kripken — COMMENTED** [review]
I don't have a deep understanding of the TLS side of things here. But looks good otherwise at a high level, aside from comments.

File: src/library_pthread.js (hunk 2 above)
Comment (sbc100) [reply src/library_pthread.js:515]: Yes, this is intended.  `PThread.tlsInitFunctions` is for future threads, and the direct call to `tlsInitWrapper` is for the currently running thread.

The previous code did this too, just in different place (library_dylink.js):

```
-        PThread.tlsInitFunctions.push(initTLS);
-        if (runtimeInitialized) {
-          initTLS();
-        }
```

Comment (sbc100) [reply src/library_pthread.js:515]: If the runtime is not yet initialized then that means the main thread has not yet called `threadInit` and in that case all the `PThread.tlsInitFunctions` will be called when it eventually does:

https://github.com/emscripten-core/emscripten/blob/8770a96c2a7a612a0e37e00370a1b707766c50e8/src/library_pthread.js#L224-L236

Comment (sbc100) [reply src/library_pthread.js:515]: In this case the check for `runtimeInitialized` is like checking for: "has current/main thread processed PThread.tlsInitFunctions already".   I'll add a comment as such.

**Review by kripken — APPROVED** [review]

File: src/library_pthread.js [hunk 3]
```diff
@@ -490,6 +490,40 @@ var LibraryPThread = {
...
+      relocateExports(metadata.tlsExports, __tls_base, /*replace=*/true);
+    }
+
+    // Register this function so that its gets called for each thread on
+    // startup.
+    PThread.tlsInitFunctions.push(tlsInitWrapper);
+
+    // If the main thread is already running we also need to call this function
+    // now.  If the main thread is not yet running this will happen when it
+    // is initializes and processes `PThread.tlsInitFunctions`.
```
Comment (dschuff) [inline src/library_pthread.js:518]:
```suggestion
    // is initialized and processes `PThread.tlsInitFunctions`.
```

**Review by dschuff — APPROVED** [review]

**Pull request merged**